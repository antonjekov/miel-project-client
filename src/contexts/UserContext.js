import React from "react";

const UserContext = React.createContext({name: "Anton"});

export default UserContext