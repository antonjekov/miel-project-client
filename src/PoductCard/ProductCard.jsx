import React from "react";
import { Card, Button } from 'react-bootstrap';
import styles from './ProductCard.module.css';
import productService from "../services/product_service";
import { useAuth } from "../contexts/Auth";
import userService from "../services/user_service";

function ProductCard(props) {

    const {userInfo} = useAuth();
    const role =userInfo? userInfo.role:''
    const {availability,_id,name,description,price} = props.product
    
    async function deleteProduct(){
        productService.delete(_id);
        window.location.reload(false);    
    }


    async function addToShoppingCard(){
        userService.shoppingCardAdd({userId:userInfo._id,productId:_id})
    }

    return (
        <Card border="warning" style={{ width: '18rem' , height: '30rem'}}>
            <Card.Img variant="top" style={{ width: '10rem', height: '10rem' }} src={props.product.imageUrl} alt='image of product' />
            <Card.Body>
                <Card.Title>{name}</Card.Title>
                <Card.Text>
                    {description}
                </Card.Text>
                <Card.Text>
                    {availability}
                </Card.Text>
                {availability.toLowerCase()==='available'?<Card.Text className={styles.Price}>
                    {`${price?.toFixed(2)} â‚¬`}
                </Card.Text>:''}
                
                {role==='admin'?<Button variant="danger" onClick={deleteProduct}>Delete product</Button>:''}
                {role!=='admin'&&availability.toLowerCase()==='available'?<Button variant="warning" onClick={addToShoppingCard} >Add to Shopping Card</Button>:''}
            </Card.Body>
        </Card>
    );
}

export default ProductCard;