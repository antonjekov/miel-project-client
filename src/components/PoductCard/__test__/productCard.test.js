import React from "react";
import ReactDOM from "react-dom"
import {render, cleanup, screen} from "@testing-library/react"
import renderer from "react-test-renderer"
import ProductCard from "../"
import { useAuth, AuthContext } from  "../../../contexts/Auth";

const productAvailable = {"_id":"5f073f0c3af80439e4f51c2c","name":"shampoo","category":"5edc2404d1fa4095ec7e954b","subcategory":"5f05703fee3bca23e884a445","price":5,"availability":"Available","imageUrl":"https://res.cloudinary.com/mielproject/image/upload/v1594310410/pfgki1klwc77ks9wfb5e.jpg"}

const productNotAvailable = {"_id":"5f0742763af80439e4f51c2e","name":"shampoo not available","category":"5edc2404d1fa4095ec7e954b","subcategory":"5f05703fee3bca23e884a445","price":6,"availability":"Not available","imageUrl":"https://res.cloudinary.com/mielproject/image/upload/v1594311284/kudbcmswwf7t03a7lht0.jpg"}


const userAdmin = {"_id":"5f073d023af80439e4f51c2a","name":"admin","username":"admin@yahoo.com","role":"admin","shoppingCard":[]}

const userClient = {"_id":"5f073d613af80439e4f51c2b","name":"anton","username":"anton@yahoo.com","role":"client","shoppingCard":["5f073f0c3af80439e4f51c2c","5f073f0c3af80439e4f51c2c"]}

const setUserInfo =()=>{}

afterEach(cleanup)

it('renders without crashing', ()=>{
    const div = document.createElement('div');
    ReactDOM.render(<AuthContext.Provider value={{ userInfo:userAdmin, setUserInfo}}>
                            <ProductCard product={productAvailable}/>
                        </AuthContext.Provider>                    
                    ,div)
})

test("displays correct button 'Delete product'", () =>{
    render(<AuthContext.Provider value={{ userInfo:userAdmin, setUserInfo}}>
        <ProductCard product={productAvailable}/>
    </AuthContext.Provider>
    )
    expect(screen.getByText("Delete product")).toBeVisible()
    const addButton = screen.queryByText("Add to Shopping Card")
    expect(addButton).not.toBeInTheDocument()
  })

  test("displays correct button 'Add to Shopping Card'", () =>{
    render(<AuthContext.Provider value={{ userInfo:userClient, setUserInfo}}>
        <ProductCard product={productAvailable}/>
    </AuthContext.Provider>
    )
    expect(screen.getByText("Add to Shopping Card")).toBeVisible()
    const deleteButton = screen.queryByText("Delete product")
    expect(deleteButton).not.toBeInTheDocument()
  })

test("not display buttons", () =>{
    render(<AuthContext.Provider value={{ userInfo:userClient, setUserInfo}}>
        <ProductCard product={productNotAvailable}/>
    </AuthContext.Provider>
    )
    const deleteButton = screen.queryByText("Delete product")
    expect(deleteButton).not.toBeInTheDocument()
    const addButton = screen.queryByText("Add to Shopping Card")
    expect(addButton).not.toBeInTheDocument()
  })

//Snapshot tests
it('matches snapshot admin', ()=>{
    const tree = renderer.create(
        <AuthContext.Provider value={{ userInfo:userAdmin, setUserInfo}}>
            <ProductCard product={productAvailable}/>
        </AuthContext.Provider>
   ).toJSON();

    expect(tree).toMatchSnapshot();
})

it('matches snapshot client product available', ()=>{
    const tree = renderer.create(
        <AuthContext.Provider value={{ userInfo:userClient, setUserInfo}}>
            <ProductCard key={productAvailable._id} product={productAvailable}/>
        </AuthContext.Provider>
   ).toJSON();
    expect(tree).toMatchSnapshot();
})

it('matches snapshot client product notAvailable', ()=>{
    const tree = renderer.create(
        <AuthContext.Provider value={{ userInfo:userClient, setUserInfo}}>
            <ProductCard key={productNotAvailable._id} product={productNotAvailable}/>
        </AuthContext.Provider>
   ).toJSON();
    expect(tree).toMatchSnapshot();
})