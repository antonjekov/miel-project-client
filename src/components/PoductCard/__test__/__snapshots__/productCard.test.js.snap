// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot admin 1`] = `
<div
  className="card border-warning"
  style={
    Object {
      "height": "25rem",
      "width": "18rem",
    }
  }
>
  
  <img
    alt="image of product"
    className="card-img-top CardImg"
    src="https://res.cloudinary.com/mielproject/image/upload/v1594310410/pfgki1klwc77ks9wfb5e.jpg"
  />
  <div
    className="card-body"
  >
    <div
      className="card-title h5"
    >
      shampoo
    </div>
    <p
      className="card-text"
    >
      Available
    </p>
    <p
      className="card-text"
    >
      <span
        className="Price"
      >
        5.00 €
      </span>
    </p>
    <div
      className="Buttons"
    >
      <button
        className="DeleteButton btn btn-danger"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
      <span>
        <button
          className="btn btn-warning"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Edit
        </button>
      </span>
    </div>
    
  </div>
</div>
`;

exports[`matches snapshot client product available 1`] = `
<div
  className="card border-warning"
  style={
    Object {
      "height": "25rem",
      "width": "18rem",
    }
  }
>
  
  <img
    alt="image of product"
    className="card-img-top CardImg"
    src="https://res.cloudinary.com/mielproject/image/upload/v1594310410/pfgki1klwc77ks9wfb5e.jpg"
  />
  <div
    className="card-body"
  >
    <div
      className="card-title h5"
    >
      shampoo
    </div>
    <p
      className="card-text"
    >
      Available
    </p>
    <p
      className="card-text"
    >
      <span
        className="Price"
      >
        5.00 €
      </span>
    </p>
    
    <button
      className="Buttons btn btn-warning"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Add to Shopping Card
    </button>
  </div>
</div>
`;

exports[`matches snapshot client product notAvailable 1`] = `
<div
  className="card border-warning"
  style={
    Object {
      "height": "25rem",
      "width": "18rem",
    }
  }
>
  <img
    alt="image of product"
    className="card-img-top CardImg"
    src="https://res.cloudinary.com/mielproject/image/upload/v1594311284/kudbcmswwf7t03a7lht0.jpg"
  />
  <div
    className="card-body"
  >
    <div
      className="card-title h5"
    >
      shampoo not available
    </div>
    <p
      className="card-text"
    >
      Not available
    </p>
    <p
      className="card-text"
    >
      
    </p>
    
    
  </div>
</div>
`;
